src/server/oauth/provider.ts

```diff
@@ -0,0 +1,716 @@
+          accessToken: tableauAccessToken,
+          refreshToken: tableauRefreshToken,
+        };
+      } else {
+        const mcpAccessToken = mcpAccessTokenSubOnlySchema.safeParse(payload);
+        if (!mcpAccessToken.success) {
+          return Err(
+            `Invalid access token: ${mcpAccessToken.error.errors.map((e) => e.message).join(', ')}`,
+          );
+        }
+
+        authInfo = {
+          userId: mcpAccessToken.data.sub,
+        };
+      }
+
+      return Ok({
+        token,
+        clientId: 'mcp-client',
+        scopes: ['read'],
+        expiresAt: payload.exp,
+        extra: authInfo,
+      });
+    } catch {
+      // TODO: Auto-refresh logic would go here
+      throw new Error('Invalid or expired access token');
+    }
+  }
+
+  /**
+   * Creates JWT access token containing credentials
+   *
+   * @remarks
+   * Part of MCP OAuth Step 7: Token Exchange
+   * JWT contains tokens for making API calls
+   *
+   * @param tokenData - token data
+   * @returns Signed JWT token for MCP authentication
+   */
+  private async createAccessToken(tokenData: UserAndTokens): Promise<string> {
+    return await new SignJWT({
+      sub: tokenData.user.name,
+      ...(this.config.auth === 'oauth'
+        ? {
+            tableauAccessToken: tokenData.tokens.accessToken,
+            tableauRefreshToken: tokenData.tokens.refreshToken,
+          }
+        : {}),
+    })
+      .setProtectedHeader({ alg: 'HS256' })
+      .setIssuedAt()
+      .setExpirationTime(Date.now() + (tokenData.tokens.expiresInSeconds - 30 * 60) * 1000) // 30 minutes before expiration
+      .setAudience(this.jwtAudience)
+      .setIssuer(this.jwtIssuer)
+      .sign(this.jwtSecret);
+  }
+
+  /**
+   * Generates PKCE code challenge from verifier
+   *
+   * @remarks
+   * Part of MCP OAuth Step 5: Authorization Request with PKCE
+   * Uses SHA256 hashing as required by S256 method
+   *
+   * @param verifier - Random code verifier string
+   * @returns Base64url-encoded SHA256 hash of verifier
+   */
+  private generateCodeChallenge(verifier: string): string {
+    return createHash('sha256').update(verifier).digest('base64url');
+  }
+
+  /**
+   * Exchanges authorization code for access tokens
+   *
+   * @remarks
+   * Part of MCP OAuth Step 6: OAuth Callback
+   * Uses token endpoint with basic auth
+   *
+   * @param code - Authorization code
+   * @param redirectUri - Redirect URI used in initial request
+   * @returns token response with access_token and refresh_token
+   */
+  private async exchangeAuthorizationCode(
+    code: string,
+    redirectUri: string,
+    clientId: string,
+    codeVerifier: string,
+  ): Promise<Result<TableauAccessToken, string>> {
+    try {
+      const result = await getTokenResult(this.config.server, {
+        grant_type: 'authorization_code',
+        code,
+        redirect_uri: redirectUri,
+        client_id: clientId,
+        code_verifier: codeVerifier,
+      });
+
+      return Ok(result);
```

Comment:
The Tableau access token is as powerful as a PAT and not scoped down to the least privileges that MCP should be afforded. Unfortunately, Tableau OAuth does not support scopes or consent UX.